Mathlib/Topology/UniformSpace/Ultra/Pseudometrizable.lean

```diff
@@ -0,0 +1,473 @@
+
+/-- Any set of pseudometrics can induce a uniform space, where the entourages are
+any open ball of positive radius for any of the pseudometrics. -/
+def UniformSpace.Core.ofPseudoMetricSystem (M : Set (PseudoMetric X ℝ)) :
+    UniformSpace.Core X where
+  uniformity := .generate <| (fun εd ↦ {xy | εd.2 xy.1 xy.2 < εd.1}) '' ((Set.Ioi 0 : Set ℝ) ×ˢ M)
+  refl := by
+    simp only [Filter.principal, idRel_subset, Filter.le_generate_iff, Set.image_subset_iff,
+      Set.preimage_setOf_eq, Set.mem_setOf_eq, PseudoMetric.refl]
+    intro
+    aesop
+  symm := by
+    rw [Filter.tendsto_iff_comap]
+    refine (Filter.generate_image_preimage_le_comap _ _).trans' ?_
+    rw [← Set.image_swap_eq_preimage_swap, Set.image_image, Set.image_swap_eq_preimage_swap]
+    simp [PseudoMetric.symm]
+  comp := by
+    rw [Filter.le_generate_iff]
+    intro s
+    simp only [Set.mem_image, Set.mem_prod, Set.mem_Ioi, Prod.exists, Filter.mem_sets,
+      forall_exists_index, and_imp]
+    rintro ε d εpos hd rfl
+    rw [Filter.mem_lift'_sets (Monotone.compRel _ _)]
+    · refine ⟨{xy | d xy.1 xy.2 < ε / 2}, Filter.mem_generate_of_mem ?_, ?_⟩
+      · simp only [Set.mem_image, Set.mem_prod, Set.mem_Ioi, Prod.exists]
+        exact ⟨ε / 2, d, ⟨by simp [εpos], hd⟩, rfl⟩
+      · intro ⟨a, b⟩
+        rw [mem_compRel]
+        simp only [Set.mem_setOf_eq, forall_exists_index, and_imp]
+        intro c hac hcb
+        refine (d.triangle _ _ _).trans_lt ((add_lt_add hac hcb).trans_le ?_)
+        simp
+    · exact monotone_id
+    · exact monotone_id
+
+/-- For the uniform space induced by a family of pseudometrics, the uniform space is
+nonarchimedean if all the pseudometrics are nonarchimedean. -/
+lemma IsUltraUniformity.ofCore_ofPseudoMetricSystem_of_isUltra {M : Set (PseudoMetric X ℝ)}
+    (hM : ∀ d ∈ M, d.IsUltra) :
+    @IsUltraUniformity _ (.ofCore <| .ofPseudoMetricSystem M) := by
+  letI : UniformSpace X := .ofCore <| .ofPseudoMetricSystem M
+  refine .mk_of_hasBasis (Filter.hasBasis_generate _) ?_ ?_
+  · intro s
+    simp only [Set.subset_image_iff, and_imp, forall_exists_index]
+    rintro hs s hs' rfl
+    refine .sInter ?_
+    simp only [Set.mem_image]
+    rintro _ ⟨⟨ε, d⟩, _, rfl⟩
+    exact d.isSymmetricRel_ball
+  · intro s
+    simp only [Set.subset_image_iff, and_imp, forall_exists_index]
+    rintro hs s hs' rfl
+    refine .sInter ?_
+    simp only [Set.mem_image]
+    rintro _ ⟨⟨ε, d⟩, hd, rfl⟩
+    exact (hM _ (hs' hd).right).isTransitiveRel_ball
+
+namespace UniformSpace.pseudoMetrizable
+
+variable [UniformSpace X]
+
+/-- A chain of nested equivalence relation entourages in a uniform space that can used to construct
+a pseudometric. -/
+structure descChainEquivRel (D : ℕ → Set (X × X)) : Prop where
+  top : D 0 = Set.univ
+  chain {n} : D (n + 1) ⊆ D n
+  mem_uniformity' {n : ℕ} (hn : n > 0) : D n ∈ 𝓤 X
+  isSymmetricRel' {n : ℕ} (hn : n > 0) : IsSymmetricRel (D n)
+  isTransitiveRel' {n : ℕ} (hn : n > 0) : IsTransitiveRel (D n)
```

Comment:
This sounds similar to the proof [UniformSpace.metrizable_uniformity](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Metrizable/Uniformity.html#UniformSpace.metrizable_uniformity), did you compare the two arguments ? Is there any factorization possibility ?